adapters/repos/db/vector/hnsw/index.go

```diff
@@ -744,3 +717,97 @@ func IsHNSWIndex(index any) bool {
 	_, ok := index.(*hnsw)
 	return ok
 }
+
+type pair struct {
+	nodeId uint64
+	level  int
+}
+
+func (h *hnsw) calculateUnreachablePoints() []uint64 {
+	h.RLock()
+	defer h.RUnlock()
+
+	visitedPairs := make(map[pair]bool)
+	candidateList := []pair{{h.entryPointID, h.currentMaximumLayer}}
+
+	for len(candidateList) > 0 {
+		currentNode := candidateList[len(candidateList)-1]
+		candidateList = candidateList[:len(candidateList)-1]
+		if !visitedPairs[currentNode] {
+			visitedPairs[currentNode] = true
+			neighbors := h.nodes[currentNode.nodeId].connectionsAtLowerLevelsNoLock(currentNode.level, visitedPairs)
+			candidateList = append(candidateList, neighbors...)
+		}
+	}
+
+	visitedNodes := make(map[uint64]bool, len(visitedPairs))
+	for k, v := range visitedPairs {
+		if v {
+			visitedNodes[k.nodeId] = true
+		}
+	}
+
+	unvisitedNodes := []uint64{}
+	for i := 0; i < len(h.nodes); i++ {
+		var id uint64
+		h.shardedNodeLocks.RLock(uint64(i))
+		if h.nodes[i] != nil {
+			id = h.nodes[i].id
+		}
+		h.shardedNodeLocks.RUnlock(uint64(i))
+		if id == 0 {
+			continue
+		}
+		if !visitedNodes[uint64(i)] {
+			unvisitedNodes = append(unvisitedNodes, id)
+		}
+
+	}
+	return unvisitedNodes
+}
+
+type HnswStats struct {
+	Dimensions         int32        `json:"dimensions"`
+	EntryPointID       uint64       `json:"entryPointID"`
+	DistributionLayers map[int]uint `json:"distributionLayers"`
+	UnreachablePoints  []uint64     `json:"unreachablePoints"`
+	NumTombstones      int          `json:"numTombstones"`
+	CacheSize          int32        `json:"cacheSize"`
+	PQConfiguration    ent.PQConfig `json:"pqConfiguration"`
+}
+
+func (s *HnswStats) IndexType() common.IndexType {
+	return common.IndexTypeHNSW
+}
+
+func (h *hnsw) Stats() (common.IndexStats, error) {
+	distributionLayers := map[int]uint{}
+
+	for _, node := range h.nodes {
```

Comment:
@asdine  Should I add the lock on the graph and on vertexes? Because I'm counting the number of nodes per layer

Reply:
Yes absolutely, you need a lock on both here